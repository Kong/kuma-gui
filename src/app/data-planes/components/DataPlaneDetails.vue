<template>
  <TabsWidget
    :tabs="filteredTabs"
    initial-tab-override="overview"
  >
    <template #tabHeader>
      <div>
        <h3>
          DPP: {{ dataPlane.name }}
        </h3>
      </div>

      <div>
        <EntityURLControl
          :name="dataPlane.name"
          :mesh="dataPlane.mesh"
        />
      </div>
    </template>

    <template #overview>
      <LabelList>
        <div>
          <ul>
            <li
              v-for="(value, prop) in processedDataPlane"
              :key="prop"
            >
              <h4>{{ prop }}</h4>

              <div v-if="prop === 'status' && typeof value !== 'string'">
                <div
                  class="entity-status"
                  :class="{
                    'is-offline': value.status.toLowerCase() === 'offline',
                    'is-online': value.status.toLowerCase() === 'online',
                    'is-degraded': value.status.toLowerCase() === 'partially degraded',
                    'is-not-available': value.status.toLowerCase() === 'not available',
                  }"
                >
                  <span>{{ value.status }}</span>
                </div>

                <div
                  v-for="(reason, index) in value.reason"
                  :key="index"
                  class="reason"
                >
                  {{ reason }}
                </div>
              </div>

              <div v-else>
                {{ value }}
              </div>
            </li>
          </ul>
        </div>

        <div>
          <template v-if="dataPlaneTags.length > 0">
            <h4>Tags</h4>

            <p>
              <span
                v-for="(tag, index) in dataPlaneTags"
                :key="index"
                class="tag-cols"
              >
                <span>
                  {{ tag.label }}:
                </span>

                <span>
                  {{ tag.value }}
                </span>
              </span>
            </p>
          </template>

          <template v-if="dataPlaneVersions">
            <h4>Versions</h4>

            <p>
              <span
                v-for="(version, dependency) in dataPlaneVersions"
                :key="dependency"
                class="tag-cols"
              >
                <span>
                  {{ dependency }}:
                </span>

                <span>
                  {{ version }}
                </span>
              </span>
            </p>
          </template>
        </div>
      </LabelList>

      <div class="config-wrapper">
        <YamlView
          id="code-block-data-plane"
          :content="rawDataPlane"
          is-searchable
        />
      </div>
    </template>

    <template #insights>
      <StatusInfo :is-empty="insightSubscriptions.length === 0">
        <KCard border-variant="noBorder">
          <template #body>
            <AccordionList :initially-open="0">
              <AccordionItem
                v-for="(insight, key) in insightSubscriptions"
                :key="key"
              >
                <template #accordion-header>
                  <SubscriptionHeader :details="insight" />
                </template>

                <template #accordion-content>
                  <SubscriptionDetails
                    :details="insight"
                    is-discovery-subscription
                  />
                </template>
              </AccordionItem>
            </AccordionList>
          </template>
        </KCard>
      </StatusInfo>
    </template>

    <template #dpp-policies>
      <DataplanePolicies
        :mesh="dataPlane.mesh"
        :dpp-name="dataPlane.name"
      />
    </template>

    <template #xds-configuration>
      <EnvoyData
        data-path="xds"
        :mesh="dataPlane.mesh"
        :dpp-name="dataPlane.name"
        query-key="envoy-data-data-plane"
      />
    </template>

    <template #envoy-stats>
      <EnvoyData
        data-path="stats"
        :mesh="dataPlane.mesh"
        :dpp-name="dataPlane.name"
        query-key="envoy-data-data-plane"
      />
    </template>

    <template #envoy-clusters>
      <EnvoyData
        data-path="clusters"
        :mesh="dataPlane.mesh"
        :dpp-name="dataPlane.name"
        query-key="envoy-data-data-plane"
      />
    </template>

    <template #mtls>
      <LabelList>
        <ul v-if="mtlsData !== null">
          <li
            v-for="(val, key) in mtlsData"
            :key="key"
          >
            <h4>{{ val.label }}</h4>

            <p>
              {{ val.value }}
            </p>
          </li>
        </ul>

        <KAlert
          v-else
          appearance="danger"
        >
          <template #alertMessage>
            This data plane proxy does not yet have mTLS configured â€”
            <a
              :href="`https://kuma.io/docs/${kumaDocsVersion}/documentation/security/#certificates`"
              class="external-link"
              target="_blank"
            >
              Learn About Certificates in {{ PRODUCT_NAME }}
            </a>
          </template>
        </KAlert>
      </LabelList>
    </template>

    <template #warnings>
      <WarningsWidget :warnings="warnings" />
    </template>
  </TabsWidget>
</template>

<script lang="ts" setup>
import { computed, ref, PropType } from 'vue'
import { KCard, KAlert } from '@kong/kongponents'

import {
  Compatibility,
  DataPlane,
  DataPlaneOverview,
} from '@/types'
import {
  compatibilityKind,
  COMPATIBLE,
  dpTags,
  getStatus,
  getVersions,
  INCOMPATIBLE_WRONG_FORMAT,
  INCOMPATIBLE_ZONE_CP_AND_KUMA_DP_VERSIONS,
  parseMTLSData,
} from '@/dataplane'
import { KUMA_ZONE_TAG_NAME, PRODUCT_NAME } from '@/consts'
import { stripTimes } from '@/helpers'
import { useStore } from '@/store/store'
import AccordionItem from '@/components/Accordion/AccordionItem.vue'
import AccordionList from '@/components/Accordion/AccordionList.vue'
import DataplanePolicies from '@/components/DataplanePolicies/DataplanePolicies.vue'
import EntityURLControl from '@/components/Utils/EntityURLControl.vue'
import EnvoyData from '@/components/EnvoyData/EnvoyData.vue'
import LabelList from '@/components/Utils/LabelList.vue'
import StatusInfo from '@/components/Utils/StatusInfo.vue'
import SubscriptionDetails from '@/views/Entities/components/SubscriptionDetails.vue'
import SubscriptionHeader from '@/views/Entities/components/SubscriptionHeader.vue'
import TabsWidget from '@/components/Utils/TabsWidget.vue'
import WarningsWidget from '@/views/Entities/components/WarningsWidget.vue'
import YamlView from '@/app/common/YamlView.vue'

const store = useStore()

const props = defineProps({
  dataPlane: {
    type: Object as PropType<DataPlane>,
    required: true,
  },

  dataPlaneOverview: {
    type: Object as PropType<DataPlaneOverview>,
    required: true,
  },
})

const tabs = [
  {
    hash: '#overview',
    title: 'Overview',
  },
  {
    hash: '#insights',
    title: 'DPP Insights',
  },
  {
    hash: '#dpp-policies',
    title: 'Policies',
  },
  {
    hash: '#xds-configuration',
    title: 'XDS Configuration',
  },
  {
    hash: '#envoy-stats',
    title: 'Stats',
  },
  {
    hash: '#envoy-clusters',
    title: 'Clusters',
  },
  {
    hash: '#mtls',
    title: 'Certificate Insights',
  },
  {
    hash: '#warnings',
    title: 'Warnings',
  },
]

const warnings = ref<Compatibility[]>([])

const processedDataPlane = computed(() => {
  const { type, name, mesh } = props.dataPlane
  const status = getStatus(props.dataPlane, props.dataPlaneOverview.dataplaneInsight)

  return {
    type,
    name,
    mesh,
    status,
  }
})

const dataPlaneTags = computed(() => dpTags(props.dataPlane))
const dataPlaneVersions = computed(() => getVersions(props.dataPlaneOverview.dataplaneInsight))
const rawDataPlane = computed(() => stripTimes(props.dataPlane))
const mtlsData = computed(() => parseMTLSData(props.dataPlaneOverview))
const insightSubscriptions = computed(() => {
  const subscriptions = Array.from(props.dataPlaneOverview.dataplaneInsight.subscriptions)

  subscriptions.reverse()

  return subscriptions
})

const kumaDocsVersion = computed(() => {
  const storedVersion = store.getters.getKumaDocsVersion

  return storedVersion !== null ? storedVersion : 'latest'
})

const filteredTabs = computed(() => warnings.value.length === 0 ? tabs.filter((tab) => tab.hash !== '#warnings') : tabs)

function setWarnings() {
  const subscriptions = props.dataPlaneOverview.dataplaneInsight.subscriptions

  if (subscriptions.length === 0 || !('version' in subscriptions[0])) {
    return
  }

  const version = subscriptions[0].version

  if (version && version.kumaDp && version.envoy) {
    const compatibility = compatibilityKind(version)

    if (compatibility.kind !== COMPATIBLE && compatibility.kind !== INCOMPATIBLE_WRONG_FORMAT) {
      warnings.value.push(compatibility)
    }
  }

  const isMulticluster = store.getters['config/getMulticlusterStatus']

  if (isMulticluster && version) {
    const tags = dpTags(props.dataPlane)
    const zoneTag = tags.find(tag => tag.label === KUMA_ZONE_TAG_NAME)

    if (zoneTag && typeof version.kumaDp.kumaCpCompatible === 'boolean' && !version.kumaDp.kumaCpCompatible) {
      warnings.value.push({
        kind: INCOMPATIBLE_ZONE_CP_AND_KUMA_DP_VERSIONS,
        payload: {
          kumaDp: version.kumaDp.version,
        },
      })
    }
  }
}

setWarnings()
</script>

<style lang="scss" scoped>
.reason {
  margin-left: var(--spacing-md);
  margin-bottom: var(--spacing-xxs);
  margin-top: var(--spacing-xxs);
}

.config-wrapper {
  padding-right: var(--spacing-md);
  padding-left: var(--spacing-md);
  padding-bottom: var(--spacing-md);
}
</style>
