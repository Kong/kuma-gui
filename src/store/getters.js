export const workspace = (state) => state.workspace

export const getSelectedMesh = (state) => {
  return state.selectedMesh
}
