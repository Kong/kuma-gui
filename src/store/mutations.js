export default {
  FETCH_MESHES (state, meshes) {
    state.meshes = meshes
  }
}
