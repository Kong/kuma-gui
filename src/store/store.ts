import { InjectionKey } from 'vue'
import { createStore, Store } from 'vuex'

import { storeConfig, State } from './index'

export const storeKey: InjectionKey<Store<State>> = Symbol('store')

export const store = createStore<State>(storeConfig)
