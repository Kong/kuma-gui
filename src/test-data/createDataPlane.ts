import { DataPlane } from '@/types'

export function createDataPlane(): DataPlane {
  return {
    type: 'Dataplane',
    mesh: 'test-mesh',
    name: 'backend',
    creationTime: '2021-02-17T08:33:36.442044+01:00',
    modificationTime: '2021-02-17T08:33:36.442044+01:00',
    networking: {
      address: '127.0.0.1',
      inbound: [
        {
          port: 7776,
          servicePort: 7777,
          serviceAddress: '127.0.0.1',
          tags: {
            'kuma.io/protocol': 'http',
            'kuma.io/service': 'backend',
          },
        },
      ],
      outbound: [
        {
          port: 10001,
          tags: {
            'kuma.io/service': 'frontend',
          },
        },
      ],
    },
  }
}
