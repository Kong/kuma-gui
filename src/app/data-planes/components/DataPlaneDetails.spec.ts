import { flushPromises, mount } from '@vue/test-utils'

import DataPlaneDetails from './DataPlaneDetails.vue'
import { store } from '@/store/store'
import { createDataPlane } from '@/test-data/createDataPlane'
import { createDataPlaneOverview } from '@/test-data/createDataPlaneOverview'

const dataPlane = createDataPlane()
const dataPlaneOverview = createDataPlaneOverview()

async function renderComponent(props = {}) {
  await store.dispatch('fetchPolicies')

  return mount(DataPlaneDetails, {
    props: {
      dataPlane,
      dataPlaneOverview,
      ...props,
    },
  })
}

describe('DataPlaneDetails', () => {
  test('matches snapshot', async () => {
    const wrapper = await renderComponent()

    await flushPromises()

    expect(wrapper.element).toMatchSnapshot()
  })

  test('has expected content and UI elements', async () => {
    const expectedOverviewStrings = ['DPP: backend', 'test-mesh', 'Online', 'kuma.io/service: backend', 'kuma.io/protocol: http', 'envoy', '1.16.2', 'kumaDp', '1.0.7', 'coredns', '1.8.3']
    const expectedInsightStrings = ['Connect time: February 17, 2021 at 7:33:36 AM', 'Control Plane Instance ID', 'Disconnect time: February 17, 2021 at 7:33:36 AM']
    const expectedPoliciesStrings = ['web', 'inbound 192.168.0.1:80:81', 'backend', 'outbound 192.168.0.2:8080']

    const wrapper = await renderComponent()

    let html = wrapper.html()
    for (const string of expectedOverviewStrings) {
      expect(html).toContain(string)
    }

    await wrapper.find('#insights-tab').trigger('click')

    html = wrapper.html()
    for (const string of expectedInsightStrings) {
      expect(html).toContain(string)
    }

    await wrapper.find('#dpp-policies-tab').trigger('click')

    // Justification: the underlying DataplanePolicies component loads additional data on mount.
    await flushPromises()

    html = wrapper.html()
    for (const string of expectedPoliciesStrings) {
      expect(html).toContain(string)
    }
  })
})
