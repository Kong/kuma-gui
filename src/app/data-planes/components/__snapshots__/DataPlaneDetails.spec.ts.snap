// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataPlaneDetails matches snapshot 1`] = `
<div
  class="tab-container"
  data-testid="tab-container"
>
  <header
    class="tab__header"
  >
    
    <div>
      <h3>
         DPP: backend
      </h3>
    </div>
    <div>
      <div
        data-testid="entity-url-control"
      >
        
        <div
          aria-controls="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
          aria-expanded="false"
          data-v-9bbde9b2=""
          id="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
          role="button"
        >
          
          <button
            class="small rounded outline k-button"
            data-v-6bcbf05d=""
            type="button"
          >
            
            <span
              class="k-button-icon kong-icon-copy kong-icon"
              data-v-637de6fa=""
              data-v-6bcbf05d=""
            >
              <svg
                height="16"
                viewBox="0 0 32 32"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                
  
                <title>
                  Copy
                </title>
                
  
                <path
                  d="M25.313 28v-18.688h-14.625v18.688h14.625zM25.313 6.688c1.438 0 2.688 1.188 2.688 2.625v18.688c0 1.438-1.25 2.688-2.688 2.688h-14.625c-1.438 0-2.688-1.25-2.688-2.688v-18.688c0-1.438 1.25-2.625 2.688-2.625h14.625zM21.313 1.313v2.688h-16v18.688h-2.625v-18.688c0-1.438 1.188-2.688 2.625-2.688h16z"
                  fill="#A3BBCC"
                />
                

              </svg>
              

            </span>
            
            
             Copy URL 
            
            <!---->
          </button>
          
          <transition-stub
            appear="false"
            css="true"
            data-v-9bbde9b2=""
            name="fade"
            persisted="false"
          >
            <div
              class="k-popover"
              data-v-9bbde9b2=""
              id="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
              role="region"
              style="width: 200px; max-width: 350px; max-height: auto; display: none;"
            >
              <!---->
              <div
                class="k-popover-content"
                data-v-9bbde9b2=""
              >
                
                <p>
                  URL copied to clipboard!
                </p>
                
              </div>
              <!---->
            </div>
          </transition-stub>
        </div>
        
      </div>
    </div>
    
  </header>
  <div
    class="tab__content-container"
  >
    <div
      class="k-tabs"
      data-v-6bcd6661=""
    >
      <ul
        aria-label="ktabs"
        data-v-6bcd6661=""
        role="tablist"
      >
        
        <li
          aria-controls="panel-0"
          aria-selected="true"
          class="active tab-item"
          data-v-6bcd6661=""
          id="overview-tab"
          role="tab"
          tabindex="0"
        >
          <a
            class="tab-link"
            data-v-6bcd6661=""
          >
            
            Overview
            
          </a>
        </li>
        <li
          aria-controls="panel-1"
          aria-selected="false"
          class="tab-item"
          data-v-6bcd6661=""
          id="insights-tab"
          role="tab"
          tabindex="0"
        >
          <a
            class="tab-link"
            data-v-6bcd6661=""
          >
            
            DPP Insights
            
          </a>
        </li>
        <li
          aria-controls="panel-2"
          aria-selected="false"
          class="tab-item"
          data-v-6bcd6661=""
          id="dpp-policies-tab"
          role="tab"
          tabindex="0"
        >
          <a
            class="tab-link"
            data-v-6bcd6661=""
          >
            
            Policies
            
          </a>
        </li>
        <li
          aria-controls="panel-3"
          aria-selected="false"
          class="tab-item"
          data-v-6bcd6661=""
          id="xds-configuration-tab"
          role="tab"
          tabindex="0"
        >
          <a
            class="tab-link"
            data-v-6bcd6661=""
          >
            
            XDS Configuration
            
          </a>
        </li>
        <li
          aria-controls="panel-4"
          aria-selected="false"
          class="tab-item"
          data-v-6bcd6661=""
          id="envoy-stats-tab"
          role="tab"
          tabindex="0"
        >
          <a
            class="tab-link"
            data-v-6bcd6661=""
          >
            
            Stats
            
          </a>
        </li>
        <li
          aria-controls="panel-5"
          aria-selected="false"
          class="tab-item"
          data-v-6bcd6661=""
          id="envoy-clusters-tab"
          role="tab"
          tabindex="0"
        >
          <a
            class="tab-link"
            data-v-6bcd6661=""
          >
            
            Clusters
            
          </a>
        </li>
        <li
          aria-controls="panel-6"
          aria-selected="false"
          class="tab-item"
          data-v-6bcd6661=""
          id="mtls-tab"
          role="tab"
          tabindex="0"
        >
          <a
            class="tab-link"
            data-v-6bcd6661=""
          >
            
            Certificate Insights
            
          </a>
        </li>
        
      </ul>
      
      <div
        aria-labelledby="overview-tab"
        class="tab-container"
        data-v-6bcd6661=""
        id="panel-0"
        role="tabpanel"
        tabindex="0"
      >
        
        
        <div>
          <div
            class="label-list-content"
          >
            <section
              aria-describedby="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
              class="noBorder kong-card"
              data-v-52c90d54=""
            >
              <!---->
              <!---->
              <div
                class="k-card-content d-flex"
                data-v-52c90d54=""
              >
                <div
                  class="k-card-body"
                  data-v-52c90d54=""
                  id="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
                >
                  
                  <div
                    class="label-list__col-wrapper multi-col"
                  >
                    
                    <div>
                      <ul>
                        
                        <li>
                          <h4>
                            type
                          </h4>
                          <div>
                            Dataplane
                          </div>
                        </li>
                        <li>
                          <h4>
                            name
                          </h4>
                          <div>
                            backend
                          </div>
                        </li>
                        <li>
                          <h4>
                            mesh
                          </h4>
                          <div>
                            default
                          </div>
                        </li>
                        <li>
                          <h4>
                            status
                          </h4>
                          <div>
                            <div
                              class="entity-status"
                            >
                              <span
                                class="entity-status__label"
                              >
                                Online
                              </span>
                            </div>
                            
                            
                          </div>
                        </li>
                        
                      </ul>
                    </div>
                    <div>
                      
                      <h4>
                        Tags
                      </h4>
                      <p>
                        
                        <span
                          class="tag-cols"
                        >
                          <span>
                            kuma.io/protocol: 
                          </span>
                          <span>
                            http
                          </span>
                        </span>
                        <span
                          class="tag-cols"
                        >
                          <span>
                            kuma.io/service: 
                          </span>
                          <span>
                            backend
                          </span>
                        </span>
                        
                      </p>
                      
                      
                      <h4>
                        Versions
                      </h4>
                      <p>
                        
                        <span
                          class="tag-cols"
                        >
                          <span>
                            envoy: 
                          </span>
                          <span>
                            1.16.2
                          </span>
                        </span>
                        <span
                          class="tag-cols"
                        >
                          <span>
                            kumaDp: 
                          </span>
                          <span>
                            1.0.7
                          </span>
                        </span>
                        <span
                          class="tag-cols"
                        >
                          <span>
                            coredns: 
                          </span>
                          <span>
                            1.8.3
                          </span>
                        </span>
                        
                      </p>
                      
                    </div>
                    
                  </div>
                  
                </div>
                <!---->
              </div>
            </section>
          </div>
        </div>
        <div
          class="config-wrapper"
        >
          <div
            class="yaml-view"
          >
            <div
              class="yaml-view-content"
            >
              <div
                class="k-tabs"
                data-v-6bcd6661=""
              >
                <ul
                  aria-label="ktabs"
                  data-v-6bcd6661=""
                  role="tablist"
                >
                  
                  <li
                    aria-controls="panel-0"
                    aria-selected="true"
                    class="active tab-item"
                    data-v-6bcd6661=""
                    id="universal-tab"
                    role="tab"
                    tabindex="0"
                  >
                    <a
                      class="tab-link"
                      data-v-6bcd6661=""
                    >
                      
                      Universal
                      
                    </a>
                  </li>
                  <li
                    aria-controls="panel-1"
                    aria-selected="false"
                    class="tab-item"
                    data-v-6bcd6661=""
                    id="kubernetes-tab"
                    role="tab"
                    tabindex="0"
                  >
                    <a
                      class="tab-link"
                      data-v-6bcd6661=""
                    >
                      
                      Kubernetes
                      
                    </a>
                  </li>
                  
                </ul>
                
                <div
                  aria-labelledby="universal-tab"
                  class="tab-container"
                  data-v-6bcd6661=""
                  id="panel-0"
                  role="tabpanel"
                  tabindex="0"
                >
                  
                  <div
                    class="code-block"
                    tabindex="0"
                  >
                    <div
                      class="code-block__actions"
                    >
                      <!--v-if-->
                      <!--v-if-->
                      <label
                        class="k-input-label code-search"
                        data-v-3472eb06=""
                        for="code-block-data-plane-search-input"
                      >
                        
                        <span
                          class="kutil-sr-only"
                        >
                          Highlight
                        </span>
                        <div
                          class="k-input-wrapper"
                          data-v-3018ca1a=""
                        >
                          <input
                            class="k-input-medium form-control k-input"
                            data-testid="search-input"
                            data-v-3018ca1a=""
                            id="code-block-data-plane-search-input"
                            type="text"
                          />
                          <!---->
                          <!---->
                        </div>
                        
                      </label>
                      <button
                        aria-pressed="false"
                        class="small outline k-button regexp-mode-button"
                        data-testid="regexp-mode-button"
                        data-v-6bcbf05d=""
                        title="Use regular expression (Alt+R)"
                        type="button"
                      >
                        
                        <!---->
                        
                        
                        <span
                          class="kutil-sr-only"
                        >
                          RegExp mode enabled
                        </span>
                         .* 
                        
                        <!---->
                      </button>
                      <button
                        aria-pressed="false"
                        class="small outline k-button filter-mode-button"
                        data-testid="filter-mode-button"
                        data-v-6bcbf05d=""
                        title="Filter results (Alt+F)"
                        type="button"
                      >
                        
                        <span
                          class="button-icon kong-icon-filter kong-icon"
                          data-v-637de6fa=""
                        >
                          <svg
                            height="11"
                            viewBox="0 0 14 11"
                            width="14"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            
  
                            <title>
                              Filter
                            </title>
                            
  
                            <path
                              d="M8 9v2H6V9h2zm2-3v2H4V6h6zm2-3v2H2V3h10zm2-3v2H0V0h14z"
                              fill="#A3BBCC"
                              fill-rule="evenodd"
                            />
                            

                          </svg>
                          

                        </span>
                        
                        
                        <span
                          class="kutil-sr-only"
                        >
                          Filter mode enabled
                        </span>
                        
                        <!---->
                      </button>
                      <button
                        class="small outline k-button previous-match-button"
                        data-testid="previous-match-button"
                        data-v-6bcbf05d=""
                        disabled=""
                        title="Previous match (Shift+Enter)"
                        type="button"
                      >
                        
                        <span
                          class="button-icon kong-icon-chevronUp kong-icon"
                          data-v-637de6fa=""
                        >
                          <svg
                            fill="none"
                            viewBox="0 0 20 20"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            
  
                            <title>
                              Collapse
                            </title>
                            
  
                            <path
                              d="M15 12 9.8 7l-5.2 5"
                              fill="none"
                              stroke="#A3BBCC"
                              stroke-linecap="round"
                              stroke-linejoin="round"
                              stroke-width="2"
                            />
                            

                          </svg>
                          

                        </span>
                        
                        
                        <span
                          class="kutil-sr-only"
                        >
                          Previous match
                        </span>
                        
                        <!---->
                      </button>
                      <button
                        class="small outline k-button next-match-button"
                        data-testid="next-match-button"
                        data-v-6bcbf05d=""
                        disabled=""
                        title="Next match (Enter)"
                        type="button"
                      >
                        
                        <span
                          class="button-icon kong-icon-chevronDown kong-icon"
                          data-v-637de6fa=""
                        >
                          <svg
                            fill="none"
                            viewBox="0 0 20 20"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            
  
                            <title>
                              Expand
                            </title>
                            
  
                            <path
                              d="m4.6 7 5.2 5L15 7"
                              fill="none"
                              stroke="#A3BBCC"
                              stroke-linecap="round"
                              stroke-linejoin="round"
                              stroke-width="2"
                            />
                            

                          </svg>
                          

                        </span>
                        
                        
                        <span
                          class="kutil-sr-only"
                        >
                          Next match
                        </span>
                        
                        <!---->
                      </button>
                    </div>
                    <div
                      class="code-block__content"
                    >
                      <!-- eslint-disable vue/no-v-html -->
                      <pre
                        class="highlighted-code-block"
                        data-testid="highlighted-code-block"
                      >
                        <code />
                      </pre>
                      <!-- eslint-enable vue/no-v-html -->
                      
                      <div
                        aria-controls="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
                        aria-expanded="false"
                        data-v-9bbde9b2=""
                        id="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
                        role="button"
                      >
                        
                        <button
                          class="code-block__copy-button"
                          type="button"
                        >
                          <span
                            class="kong-icon-copy kong-icon"
                            data-v-637de6fa=""
                          >
                            <svg
                              height="16"
                              viewBox="0 0 32 32"
                              width="16"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              
  
                              <title>
                                Copy
                              </title>
                              
  
                              <path
                                d="M25.313 28v-18.688h-14.625v18.688h14.625zM25.313 6.688c1.438 0 2.688 1.188 2.688 2.625v18.688c0 1.438-1.25 2.688-2.688 2.688h-14.625c-1.438 0-2.688-1.25-2.688-2.688v-18.688c0-1.438 1.25-2.625 2.688-2.625h14.625zM21.313 1.313v2.688h-16v18.688h-2.625v-18.688c0-1.438 1.188-2.688 2.625-2.688h16z"
                                fill="#A3BBCC"
                              />
                              

                            </svg>
                            

                          </span>
                          <span
                            class="kutil-sr-only"
                          >
                            Copy
                          </span>
                        </button>
                        
                        <transition-stub
                          appear="false"
                          css="true"
                          data-v-9bbde9b2=""
                          name="fade"
                          persisted="false"
                        >
                          <div
                            class="k-popover"
                            data-v-9bbde9b2=""
                            id="aaaabbbb-cccc-dddd-eeee-ffffffffffff"
                            role="region"
                            style="width: 200px; max-width: 350px; max-height: auto; display: none;"
                          >
                            <!---->
                            <div
                              class="k-popover-content"
                              data-v-9bbde9b2=""
                            >
                              
                              <p>
                                Code copied to clipboard
                              </p>
                              
                            </div>
                            <!---->
                          </div>
                        </transition-stub>
                      </div>
                      
                    </div>
                  </div>
                  
                </div>
                <div
                  aria-labelledby="kubernetes-tab"
                  class="tab-container"
                  data-v-6bcd6661=""
                  id="panel-1"
                  role="tabpanel"
                  tabindex="0"
                >
                  <!---->
                </div>
                
              </div>
            </div>
          </div>
        </div>
        
        
      </div>
      <div
        aria-labelledby="insights-tab"
        class="tab-container"
        data-v-6bcd6661=""
        id="panel-1"
        role="tabpanel"
        tabindex="0"
      >
        <!---->
      </div>
      <div
        aria-labelledby="dpp-policies-tab"
        class="tab-container"
        data-v-6bcd6661=""
        id="panel-2"
        role="tabpanel"
        tabindex="0"
      >
        <!---->
      </div>
      <div
        aria-labelledby="xds-configuration-tab"
        class="tab-container"
        data-v-6bcd6661=""
        id="panel-3"
        role="tabpanel"
        tabindex="0"
      >
        <!---->
      </div>
      <div
        aria-labelledby="envoy-stats-tab"
        class="tab-container"
        data-v-6bcd6661=""
        id="panel-4"
        role="tabpanel"
        tabindex="0"
      >
        <!---->
      </div>
      <div
        aria-labelledby="envoy-clusters-tab"
        class="tab-container"
        data-v-6bcd6661=""
        id="panel-5"
        role="tabpanel"
        tabindex="0"
      >
        <!---->
      </div>
      <div
        aria-labelledby="mtls-tab"
        class="tab-container"
        data-v-6bcd6661=""
        id="panel-6"
        role="tabpanel"
        tabindex="0"
      >
        <!---->
      </div>
      
    </div>
    <div>
      <!--v-if-->
    </div>
  </div>
</div>
`;
